sycl/source/builtins_generator.py

```diff
@@ -0,0 +1,1372 @@
+  """Returns true if the argument type is a templated multi_ptr."""
+  return isinstance(arg_type, InstantiatedTemplatedArg) and isinstance(arg_type.templated_type, MultiPtr)
+
+def convert_arg_name(arg_type, arg_name):
+  """
+  Converts an argument name to a valid argument to be passed to another
+  function. For vector arguments this is done by converting the argument into
+  the associated vector type, needed for swizzles.
+  """
+  if is_vec_arg(arg_type):
+    return f'detail::simplify_if_swizzle_t<{arg_type}>({arg_name})'
+  return arg_name
+
+def get_invoke_args(arg_types, arg_names, convert_args=[]):
+  """
+  Gets the arguments to be used inside builtins when calling the implementation
+  invoke functions.
+  """
+  result = list(map(convert_arg_name, arg_types, arg_names))
+  for (arg_idx, type_conv) in convert_args:
+    # type_conv is either an index or a conversion function/type.
+    conv = type_conv if isinstance(type_conv, str) else arg_types[type_conv]
+    if is_vec_arg(conv):
+      # If the conversion is to a vector template argument, it could also be
+      # a swizzle. Since we cannot convert most types to swizzles, we make
+      # sure to make the conversion to the corresponding vector type instead.
+      conv = f'detail::simplify_if_swizzle_t<{conv}>'
+    result[arg_idx] = f'{conv}({result[arg_idx]})'
+  return result
+
+class DefCommon:
+  """
+  Common super class representing a builtin definition.
+  return_type - The name of the type group of the return type. Must match a key
+    in builtin_types.
+  arg_types - The names of the type group of each argument in the builtin.
+  invoke_name - The implementation invoke function name associated with the
+    builtin. For scalar and vector builtins this is prefixed by
+    __sycl_std::__invoke_.
+  custom_invoke - A function object to be used for generating a custom body for
+    the builtin.
+  size_alias - A name to use as an alias for the size associated with vector and
+    marray arguments.
+  marray_use_loop - A bool specifying if marray builtins from this definition
+    should use a loop-based implementation instead of the default vectorized
+    implementation used when possible.
+  template_scalar_args - A bool specifying if the builtin should combine the
+    scalar arguments into common template types.
+  """
+  def __init__(self, return_type, arg_types, invoke_name, invoke_prefix,
+               custom_invoke, size_alias, marray_use_loop,
+               template_scalar_args):
+    self.return_type = return_type
+    self.arg_types = arg_types
+    self.invoke_name = invoke_name
+    self.invoke_prefix = invoke_prefix
+    self.custom_invoke = custom_invoke
+    self.size_alias = size_alias
+    self.marray_use_loop = marray_use_loop
+    self.template_scalar_args = template_scalar_args
+
+  def require_size_alias(self, alternative_name, marray_type):
+    """
+    Requires that a size alias is specified in the function boy. This returns
```

Comment:
```suggestion
    Requires that a size alias is specified in the function body. This returns
```

Reply:
Hah! Personally I prefer it the way it is, but I will reluctantly change it. 😜